### PR TITLE
 Add ByteSelectiveStreamReader

### PR DESCRIPTION
**Pull request by bhhari**
This commit adds a new SelectiveStreamReader for TINYINT type.
Filters the values during the read.
This PR is part 1 of the #13174

### CODE REVIEW

**Comment by bhhari** [comment]
@mbasmanova added benchmark test and also making the readWithFilter as a new method performs a bit better in this case, but not that significant. so removed the comment and kept it inline as of now.

BenchmarkTest results  (batch column has  the data for BatchStreamer)

```
Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score   batch
BenchmarkSelectiveStreamReaders.read                  tinyint         true  avgt   20    0.039    0.105      
BenchmarkSelectiveStreamReaders.read                  tinyint        false  avgt   20    0.027    0.036        
BenchmarkSelectiveStreamReaders.readAllNull           tinyint          N/A  avgt   20    0.003    0.002      



In new method

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.readWithFilter          tinyint         true  avgt   20  0.077 ± 0.004   s/op
BenchmarkSelectiveStreamReaders.readWithFilter          tinyint        false  avgt   20  0.083 ± 0.004   s/op


Inline

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.readWithFilter          tinyint         true  avgt   20  0.084 ± 0.004   s/op
BenchmarkSelectiveStreamReaders.readWithFilter          tinyint        false  avgt   20  0.081 ± 0.003   s/op
```

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 1]
```diff
@@ -50,11 +52,13 @@ private static QueryRunner createQueryRunner()
                 Optional.empty());
 
         queryRunner.execute(noPushdownFilter(queryRunner.getDefaultSession()),
-                "CREATE TABLE lineitem_ex (linenumber, orderkey, ship_by_air, is_returned) AS " +
+                "CREATE TABLE lineitem_ex (linenumber, orderkey, ship_by_air, is_returned, byte_col1, byte_col2) AS " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: How would you feel about using more descriptive column names? For example, day(shipdate) as ship_day and month(shipdate) as ship_month? These 2 will give you values within TINYINT range and will have meaningful names that are easy to understand.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 2]
```diff
@@ -84,6 +88,32 @@ public void testBooleans()
         assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE ship_by_air is not null AND is_returned = true");
     }
 
+    @Test
+    public void testBytes()
+    {
+        // Single tinyint column
+        assertQueryUsingH2Cte("SELECT byte_col1 FROM lineitem_ex");
+
+        assertQueryUsingH2Cte("SELECT byte_col1 FROM lineitem_ex WHERE byte_col1 < 0");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: This query returns no data. Is this intentional? If so, perhaps, use `assertQueryReturnsEmptyResult`.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 3]
```diff
@@ -84,6 +88,32 @@ public void testBooleans()
...
+
+        assertQueryUsingH2Cte("SELECT count(*) FROM lineitem_ex WHERE byte_col1 > 0");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: This filter doesn't filter anything. Is this intentional?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 4]
```diff
@@ -84,6 +88,32 @@ public void testBooleans()
...
+
+        assertQueryUsingH2Cte("SELECT count(*) FROM lineitem_ex WHERE byte_col1 is null");
+
+        // Two tinyint columns
+        assertQueryUsingH2Cte("SELECT byte_col1, byte_col2 FROM lineitem_ex");
+
+        assertQueryUsingH2Cte("SELECT byte_col1, byte_col2 FROM lineitem_ex WHERE byte_col1 = 0");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: No data matches this filter. Is this intentional? If so, use assertQueryReturnsEmptyResult.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 5]
```diff
@@ -84,6 +88,32 @@ public void testBooleans()
...
+
+        assertQueryUsingH2Cte("SELECT byte_col1, byte_col2 FROM lineitem_ex WHERE byte_col1 = 0 AND byte_col2 = 0");
+
+        assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE byte_col1 is null");
+
+        assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE byte_col1 is not null AND byte_col2 = 1");
+
+        assertQueryUsingH2Cte("SELECT byte_col1, byte_col2 FROM lineitem_ex WHERE byte_col1 > 0 and byte_col2 < 0 and (byte_col1 + byte_col2) = 0");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: `(byte_col1 + byte_col2) = 0` is always true here. Is this intentional? Perhaps, add another filter that excludes some rows.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 6]
```diff
@@ -84,6 +88,32 @@ public void testBooleans()
         assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE ship_by_air is not null AND is_returned = true");
     }
 
+    @Test
+    public void testBytes()
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: I'm curious how good is test coverage of this test?

Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: line coverage is 82%.
com.facebook.presto.orc.reader.ByteSelectiveStreamReader | 100% (1/1) | 100% (19/19) | 82% (200/243)
-- | -- | -- | --

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 5 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: using ship_day and ship_month as suggested above

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 4 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: using ship_day and ship_month as suggested above

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 3 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: using ship_day and ship_month as suggested above

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 2 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: using ship_day and ship_month as suggested above

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 7]
```diff
@@ -31,6 +31,8 @@
 {
     private static final String WITH_LINEITEM_EX = "WITH lineitem_ex AS (" +
             "SELECT linenumber, orderkey, " +
+            "   CASE WHEN linenumber % 9 = 0 THEN NULL ELSE cast(day(shipdate) as TINYINT) END AS ship_day, " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: make all SQL keywords uppercase, e.g. `cast(... as TINYINT)` -> `CAST(... AS TINYINT)`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 8]
```diff
@@ -50,11 +52,13 @@ private static QueryRunner createQueryRunner()
...
 
         queryRunner.execute(noPushdownFilter(queryRunner.getDefaultSession()),
-                "CREATE TABLE lineitem_ex (linenumber, orderkey, ship_by_air, is_returned) AS " +
+                "CREATE TABLE lineitem_ex (linenumber, orderkey, ship_by_air, is_returned, ship_day, ship_month) AS " +
                         "SELECT linenumber, " +
                         "   orderkey, " +
                         "   IF (linenumber % 7 = 0, null, shipmode = 'AIR') AS ship_by_air, " +
-                        "   IF (linenumber % 5 = 0, null, returnflag = 'R') AS is_returned " +
+                        "   IF (linenumber % 5 = 0, null, returnflag = 'R') AS is_returned, " +
+                        "   IF (linenumber % 4 = 0, null, CAST(day(shipdate) as TINYINT)) AS ship_day, " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: as -> AS

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 9]
```diff
@@ -50,11 +52,13 @@ private static QueryRunner createQueryRunner()
...
+                        "   IF (linenumber % 6 = 0, null, CAST(month(shipdate) as TINYINT)) AS ship_month " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: as -> AS

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 10]
```diff
@@ -84,6 +88,32 @@ public void testBooleans()
...
+
+        assertQueryUsingH2Cte("SELECT ship_day, ship_month FROM lineitem_ex WHERE ship_month = 1");
+
+        assertQueryUsingH2Cte("SELECT ship_day, ship_month FROM lineitem_ex WHERE ship_day = 1 AND ship_month = 1");
+
+        assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE ship_month is null");
+
+        assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE ship_day is not null AND ship_month = 1");
+
+        assertQueryUsingH2Cte("SELECT ship_day, ship_month FROM lineitem_ex WHERE ship_day > 15 and ship_month < 5 and (ship_day + ship_month) < 20");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: and -> AND

**Pull request merged**